### ISSUE TITLE
`Model` subclasses cannot be loaded without "losing" their class. 

### ISSUE THREAD

**Issue by jamlong** [tensorflow/tensorflow#45781]
**System information**

- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
Yes.  I have included a short, self-contained, correct example in a Colab below.

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Reproducible on Colab instances.  Originally identified on Debian 10, running on Google Cloud 
Linux jamlong-gpu-4 4.9.0-12-amd64 #1 SMP Debian 4.9.210-1 (2020-01-20) x86_64 GNU/Linux

- TensorFlow installed from (source or binary):
Binary (pip install tensorflow==2.4.0)

- TensorFlow version (use command below):
v2.4.0-rc4-71-g582c8d236cb 2.4.0

- Python version:
Python 3.7.6

- CUDA/cuDNN version:
N/A - reproducible on Colab without GPU

- GPU model and memory:
N/A - reproducible on Colab without GPU

**Describe the current behavior**
Loading a custom `MySubClass` model (subclass of `Model`) from a SavedModel yields a `Functional` instead of a `MySubClass`.

**Describe the expected behavior**
I'd expect when I define a `MySubClass` class as a subclass of Model, and then reload an instance of the model, I would get a `MySubClass`

**Standalone code to reproduce the issue**

Colab demonstrating an SSCCE:

https://colab.research.google.com/drive/12ShtOJ7oBkQsGsn_N801qHeMv3__POv4?usp=sharing

**Other info / logs** 

The issue appears to be that inheritance is effectively "broken" for `Model.from_config` as a result of the introduction of the `Functional` class, which `Model` is proxying to, in somewhat baffling (to an outsider, at least) fashion. 

If a user creates `MySubClass` as subclass of `Model`, and doesn't define a `from_config`, the implementation falls back to `Model.from_config`:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/training.py#L2302
```
 @classmethod
  def from_config(cls, config, custom_objects=None):
    # Since only FunctionalModel produces config, the model can only
    # be constructed for FunctionalModel
    from tensorflow.python.keras.engine import functional  # pylint: disable=g-import-not-at-top
    return functional.Functional.from_config(
        config, custom_objects=custom_objects)
```

However, this implementation assumes that everything is really a `Functional` (since any model with a config is a `Functional`), which is what appears to be breaking us in this case. `Model.from_config(MySubClass, config...)` gets called, and it delegates to an explicit `Functional.from_config(config...)`, discarding the fact that cls=MySubClass in the process.  

The problem is, `Model.from_config` (or, more accurately, `Functional.from_config` is doing _all of the work_ in reloading the model, so we're completely reliant on it. However, the fact that the `Model` / `Functional` differentiation exists completely breaks the ability to use it.

* It doesn't work "out of the box", which it ideally would without the `Model` / `Functional` divide.
* Normally you'd just do any pre-work, defer to `super().from_config()`, and do any post-processing. However, the super-call would trigger the "Force it to be `Functional`" code, which still breaks it.
* Because `Functional` is part of `tf.python.keras` and not exported as part of the public API, we can't simply emulate the logic of `Functional.from_config` (which makes call to `tf.python.keras.engine.functional`,  delegates to the `cls` constructor, and then another call to do some post-processing)
* Of note: if I ignore the fact that it's _also_ completely the wrong thing to do, and have my class subclass  `tf.python.keras.engine.functional.Functional` instead of `Model`, it appears to work. It looks like it may _just_ be the fact that `Model.from_config` discards the `cls`.  But again, not something we can do here since it's not part of the public API.

In short: unless I've missed something obvious, model subclassing is currently broken w.r.t. serialization because of the `Model` vs. `Functional` divide.

**Comment by amahendrakar** [comment]
@jamlong,
I do not have access to the Colab notebook you have linked. Could you please provide the required permissions to view the file. Thanks!

**Comment by jamlong** [comment]
Updated the link to something more accessible:
https://colab.research.google.com/drive/12ShtOJ7oBkQsGsn_N801qHeMv3__POv4?usp=sharing

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.3, TF v2.4 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/328db4c6cafbc39a3aeb6926764f37a3/45781.ipynb). Thanks!

**Comment by monicadsong** [comment]
Hi--

Everything you say is correct, and thank you for bringing this up-- I am looking into it. In the meantime, though, I would hesitate to say that "model subclassing is currently broken w.r.t. serialization because of the Model vs. Functional divide", since subclassed model can be revived as the correct class if you register the class as a custom object:
```
reloaded_model = tf.keras.models.load_model(tmpdir, custom_objects={"MyModelSubclass": MyModelSubclass})
assert isinstance(reloaded_model, MyModelSubclass)
```

**Comment by monicadsong** [comment]
I am interested in getting a user's perspective here, @jamlong.

Since the below works:
```
reloaded_model = tf.keras.models.load_model(tmpdir, custom_objects={"MyModelSubclass": MyModelSubclass})
assert isinstance(reloaded_model, MyModelSubclass)
```
do you still believe that model inheritance is broken?

Also, another question:
```
class MyModelSubclass(tf.keras.models.Model):
    pass
model = MyModelSubclass()
model.save(tmpdir)
del MyModelSubclass  # delete class definition since user may not have access to it in a new runtime

reloaded_model = tf.keras.models.load_model(tmpdir)
print(type(reloaded_model))
```
What do you expect the last line to print?

**Comment by jamlong** [comment]
> I would hesitate to say that "model subclassing is currently broken
w.r.t. serialization because of the Model vs. Functional divide", since
subclassed model can be revived as the correct class if you register the
class as a custom object:

Ah. The example I provided is missing an important detail.  The original
code had the subclass decorated with
@tf.keras.utils.register_keras_serializable, which we'd expect to be
sufficient based on its documentation:

This decorator injects the decorated class or function into the Keras
custom object dictionary,* so that it can be serialized and deserialized
without needing an entry in the user-provided custom object dict.*


Is this not a sufficient approach in this case for model subclasses, for
some reason?

If you update the example to use it, you still run into the same issue.  I
rewrote this as a standalone script to rule out colab issues.  Updated
Gist:
https://gist.github.com/jamlong/d1d5cf61128d11191631ca71a8825a52

If you run the gist under tf 2.4.0, the first run will define/register the
subclass, create a model, save it, and then reload it within the same
process. This is fine, and produces output like:

Reloaded model:

<__main__.MyModelSubclass object at 0x7f5f6ee2f490>

Restart the Runtime & run all

If you run it a second time from a separate process (as in: save the gist
as a script and run it twice, or "Restart Runtime And Run All" in colab),
the second time it defines the subclass and registers it, sees a model file
exists, then simply reloads the model from disk.  This still results in the
Model being reloaded as a `Functional` instead of `MySubClass` despite
MySubClass being defined and registered with the decorator.

<tensorflow.python.keras.engine.functional.Functional object at
0x7f78bb3d4ad0>

Traceback (most recent call last):

  File "test.py", line 20, in <module>

    f"Reloaded instance of type {type(reloaded_model)}, but should also be
MyModelSubclass"

AssertionError: Reloaded instance of type <class
'tensorflow.python.keras.engine.functional.Functional'>, but should also be
MyModelSubclass

> do you still believe that model inheritance is broken?

Yes-ish. I'd expect the new gist to work, and it doesn't.

At this point it's unclear to me which bit isn't working, whether that's an
issue with the inheritance (because if I derive from the non-public
`Functional` class directly instead of `Model` it works...) or the
serialization (register_keras_serializable/load_model), or my *expectations* of
register_keras_serializable working in this case.

**Comment by ds-steventondeur** [comment]
I confirm jamlong's observations. The following does not work for me; the reloaded model is of type `Functional`.
My use case: I train a custom model in one script, and I try to load the trained model in another script.

> ```
> reloaded_model = tf.keras.models.load_model(tmpdir, custom_objects={"MyModelSubclass": MyModelSubclass})
> assert isinstance(reloaded_model, MyModelSubclass)  # ASSERT --> reloaded_model = <tensorflow.python.keras.engine.functional.Functional object at 0x7f40d2908c50>
> ```

**Comment by pandrey-fr** [comment]
As a TensorFlow user, I concur with @jamlong as to the fact that the identified behavior is indeed an issue. I encounter it both with TF 2.3 and TF 2.4 (using Python 3.6 and 3.8 on various Linux-based systems).

I have implemented a custom Model (let's call it `CustomModel`) subclass to override some Functional behaviors (namely, to implement gradient accumulation and hack loss/metrics containers into supporting RaggedTensor inputs), and run into the exact same issue: in spite of it being registered (and even if I explicitly pass a `custom_objects` dict to reference it), depending on whether such a `CustomModel` has already been instantiated in the current thread, de-serialization from a save file and/or config dict will either produce a `FunctionalModel` or `CustomModel`.

This causes serious issues as depending on the restored class, the model will or will not implement behaviors that are absolutely necessary for my code to run. At the moment I am considering using a ugly hack that involves instantiating a `CustomModel` based on reloaded layers, optimizer, etc. gatherd from the `FunctionalModel` when the issue happens, but this is impractical since I would basically have to write such a hack for each and every of my custom Model classes - and I have quite a few - and would add unwanted elements to the computation graph, which is not nice for TensorBoard visualization (although this is rather minor).

**Comment by monicadsong** [comment]
Hi all--

Thanks for all your thoughtful comments.

An update:
(1) Re @jamlong 's issue, I believe it is now fixed in the latest tf-nightly, see the link here: https://colab.research.google.com/gist/monicadsong/af9b7085f97d00b94894a460daa78a33/45781-tf-nightly.ipynb

(2) @ds-steventondeur: reloading a subclass Functional model should work now in `tf-nightly` without fail.

(3) @pandrey-fr: this issue is now fixed in `tf-nightly` and will be in TF 2.5. If you would like to continue using your current version of TF without updating, you are correct in that will need to instantiate a `CustomModel` in your current thread. Maybe `tf.compat.v1.reset_default_graph` will clear these these dummy models from the computational graph.

You can also try doing something like: 
```
# Change CustomModel to inherit from `Functional`
CustomModel.__bases__ = (tensorflow.python.keras.engine.functional.Functional,)
```

I am closing this issue now, but please reopen if I overlooked something!

**Comment by pandrey-fr** [comment]
Hi @monicadsong, thank you very much for your answer!
I will try and implement your suggested fix while waiting for TF 2.5 to land :)
